### ISSUE TITLE
[BUG] Haproxy 2.8 option forwardfor if-none not working anymore

### ISSUE THREAD

**Issue by robinostlund** [haproxy/haproxy#2187]
### Detailed Description of the Problem

I have recently upgrade from haproxy 2.7 to haproxy 2.8. And it seems that "option forwardfor if-none" is not working anymore, i get multiple values in the x-forwarded-for header and i have confirmed by going back to 2.7 that it starts to work again.

### Expected Behavior

It should not add x-forwarded-for header if it already exist.

### Steps to Reproduce the Behavior

1. Upgrade to 2.8
2. Enable option forwardfor if-none

### What is your configuration?

```haproxy
# {{ ansible_managed }}
#-----------------------------------#
#             globals               #
#-----------------------------------#
global
	#*-------------------------*#
	#      chroot settings      #
	#*-------------------------*#
	chroot /usr/local/etc/haproxy

	#*-------------------------*#
	#        log settings       #
	#*-------------------------*#
	log stdout format raw local0 warning

	#*-------------------------*#
	#       stats settings      #
	#*-------------------------*#
	stats timeout 30s

	#*-------------------------*#
	#       conn settings       #
	#*-------------------------*#
	maxconn 1000


	#*-------------------------*#
	#        ssl settings       #
	#*-------------------------*#
	ca-base /etc/ssl/certs
	crt-base /etc/ssl/private
	ssl-default-bind-ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256
  	ssl-default-bind-options ssl-min-ver TLSv1.2 ssl-max-ver TLSv1.3
  	ssl-default-server-ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256
  	ssl-default-server-options ssl-min-ver TLSv1.2 ssl-max-ver TLSv1.3

	#*-------------------------*#
	#          secrets          #
	#*-------------------------*#
	#cluster-secret ebe0ohBahnoiqu3Qui9n

	#*-------------------------*#
	#			lua  			#
	#*-------------------------*#
	#lua-load /usr/local/etc/haproxy/lua/cors.lua
	lua-load /usr/local/etc/haproxy/lua/robots_disallow.lua
	lua-load /usr/local/etc/haproxy/lua/healthcheck.lua

#-----------------------------------#
#             defaults              #
#-----------------------------------#
defaults

	#*-------------------------*#
	#        misc settings      #
	#*-------------------------*#
	mode http
	option splice-auto
	option redispatch
	# never fail on address resolution
    default-server init-addr last,libc,none

	#*-------------------------*#
	#    connection settings    #
	#*-------------------------*#
	option http-server-close
	option forwarded

	#*-------------------------*#
	#        log settings       #
	#*-------------------------*#
	log	global
	option httplog
	option dontlognull
	#option dontlog-normal # don't log ok requests, only errors.

	#*-------------------------*#
	#       loadbalacing        #
	#*-------------------------*#
	balance leastconn
	#http-check expect status 200
	#http-check send-state

	#*-------------------------*#
	#          timeout          #
	#*-------------------------*#
	# https://delta.blue/blog/haproxy-timeouts/
	timeout	connect	5s
	timeout client	30s
	timeout server	30s
	timeout tunnel 20m
	timeout queue 5s
	timeout http-request 10s
	timeout http-keep-alive 2s

	#*-------------------------*#
	#        compression        #
	#*-------------------------*#
	compression algo gzip
    compression type text/css text/html text/javascript application/javascript text/plain text/xml application/json image/svg+xml

	#*-------------------------*#
	#          errors           #
	#*-------------------------*#
	errorfile 400 /usr/local/etc/haproxy/errors/400.http
	errorfile 401 /usr/local/etc/haproxy/errors/401.http
	errorfile 403 /usr/local/etc/haproxy/errors/403.http
	errorfile 404 /usr/local/etc/haproxy/errors/404.http
	errorfile 408 /usr/local/etc/haproxy/errors/408.http
	errorfile 500 /usr/local/etc/haproxy/errors/500.http
	errorfile 502 /usr/local/etc/haproxy/errors/502.http
	errorfile 503 /usr/local/etc/haproxy/errors/503.http
	errorfile 504 /usr/local/etc/haproxy/errors/504.http


#-----------------------------------#
#             resolvers             #
#-----------------------------------#
resolvers mydns
	parse-resolv-conf
	hold valid 5s

resolvers docker
	nameserver dns 127.0.0.11:53

resolvers cloudlare
	nameserver cloudflare_1 1.1.1.1:53
	nameserver cloudflare_2 1.0.0.1:53
	timeout resolve 10s

#-----------------------------------#
#             userlists             #
#-----------------------------------#
userlist admins
	user hidden password hidden

#-----------------------------------#
#           frontends               #
#-----------------------------------#
frontend prometheus-stats
   bind *:{{ docker_tcp_port_haproxy_prometheus }}
   http-request use-service prometheus-exporter if { path /metrics }
   stats enable
   stats uri /stats
   stats refresh 10s

frontend web
	#*-------------------------*#
	#            bind           #
	#*-------------------------*#
	#bind *:{{ docker_tcp_port_haproxy_http }} 		tfo
	#bind *:{{ docker_tcp_port_haproxy_https }} 		ssl crt /etc/letsencrypt/live/hidden.se/priv-fullchain-bundle.pem tfo alpn h2,http/1.1 npn h2,http/1.1
	bind *:{{ docker_tcp_port_haproxy_http }}		
	bind *:{{ docker_tcp_port_haproxy_https }} 		ssl crt /etc/letsencrypt/live/hidden.se/priv-fullchain-bundle.pem alpn h2,http/1.1

	# enables HTTP/3 over QUIC (not in docker version yet)
	bind quic4@:{{ docker_tcp_port_haproxy_https }}	ssl crt /etc/letsencrypt/live/hidden.se/priv-fullchain-bundle.pem alpn h3

	#*-------------------------*#
	#        log settings       #
	#*-------------------------*#
	log hidden:514 local0

	#*-------------------------*#
	#         log format        #
	#*-------------------------*#
	# https://www.haproxy.com/documentation/aloha/12-5/traffic-management/lb-layer7/log-format-rules/
	# default: %ci:%cp [%tr] %ft %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs %{+Q}r
	# with response times
	#log-format "[%tr] %ci:%cp->%fi:%fp %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs %{+Q}r"
	#log-format '[%tr] "%ci:%cp->%fi:%fp->%b/%s" "%ST<-%B<-%si" "%hr" "%hs" %{+Q}r'
	#log-format '[%tr] "%ST" "%ci:%cp->%b<-%si:%sp" "%hr" "%hs" %{+Q}r'
	log-format '"%[var(txn.prxy_mthd)]" "%ST" "%ci:%cp->%b<-%si:%sp" %{+Q}r "%hr"'

	#*-------------------------*#
	#          captures         #
	#*-------------------------*#
	capture request header Host len 100
	#capture request header Content-Length len 10
	#capture request header Accept-Language len 64
	capture request header Referer len 64
	capture request header X-Forwarded-For len 100
	capture request header User-Agent len 64
    capture request header CF-IPCountry len 64
    capture request header CF-Connecting-IP len 64
	capture request header X-ProxyHost len 64

	# capture response header X-Cache-Status len 100
    # capture response header Content-Length len 24
	# capture response header Location len 32


	#*-------------------------*#
	#			options         #
	#*-------------------------*#
	option forwardfor if-none 

	#*-------------------------*#
	#         acls: misc        #
	#*-------------------------*#
	# check if request is letsencrypt
	# acl letsencrypt_acl path_beg /.well-known/acme-challenge/

	# check if request is haproxy_status
	#acl status_page path_beg /haproxy

	# prevent ssl redirects for unifi
	acl unifi_no_ssl_redirect path_beg /dl /inform

	# test if request is robots.txt
	acl robots_disallow path_end -i robots.txt

	# check if HSTS exists
	# acl hsts_exists req.hdr_cnt(Srict-Transport-Security) gt 0

	# check if we have a websocket
	acl is_websocket hdr(Upgrade) -i WebSocket

	# check if we have static files
	acl cache_short_static_file capture.req.uri -m end .css .js
	acl cache_long_static_file capture.req.uri -m end .jpg .jpeg .gif .ico .png .bmp .woff .pdf

	#*-------------------------*#
	#     acls: auth     #
	#*-------------------------*#
	acl auth_required req.hdr(host),lower,map_dom(/usr/local/etc/haproxy/maps/domain_auth.map) -m found
	acl auth_required_unrestricted src,map_ip(/usr/local/etc/haproxy/maps/admin_ip_list_unrestricted.map) -m found
	acl auth_ok http_auth(admins)

	#*-------------------------*#
	#     acls: healthcheck     #
	#*-------------------------*#
	# test uri to see if its a healtcheck request
	acl healtcheck_allowed_nets src 127.0.0.1
	acl healthcheck path_beg /healthcheck


	#*-------------------------*#
	#        unique ids         #
	#*-------------------------*#
	# add unique id to requests
	unique-id-format %{+X}o\ %ci:%cp_%fi:%fp_%Ts_%rt:%pid
	unique-id-header X-Request-ID

	#*-------------------------*#
	#    request: cloudflare    #
	#*-------------------------*#
	# add which port we are connecting from
	acl is_cloudflare_ipv4 src -f /usr/local/etc/haproxy/cf-ips-v4
	acl is_cloudflare_ip_hdr req.hdr(CF-Connecting-IP) -m found
	acl is_cloudflare_country_hdr req.hdr(Cf-Ipcountry) -m found
	http-request set-var(txn.prxy_mthd) str("cloudflare") if is_cloudflare_ipv4
	http-request set-var(txn.prxy_mthd) str("tunnel") if !is_cloudflare_ipv4 is_cloudflare_ip_hdr
	http-request set-var(txn.prxy_mthd) str("direct") if !is_cloudflare_ipv4 !is_cloudflare_ip_hdr
	http-request set-header X-Country %[req.hdr(Cf-Ipcountry)] if is_cloudflare_country_hdr
	
	# must be here otherwise admin ip check wont work
	http-request set-src req.hdr(CF-Connecting-IP) if is_cloudflare_ip_hdr is_cloudflare_ipv4

	#*-------------------------*#
	#      request: services    #
	#*-------------------------*#
	http-request use-service lua.healthcheck if healthcheck healtcheck_allowed_nets	
	http-request use-service lua.robots if robots_disallow { req.hdr(host),lower,map_dom(/usr/local/etc/haproxy/maps/no_robots.map) -m found } 


	#*-------------------------*#
	#    request: set-header    #
	#*-------------------------*#
	http-request set-header X-Forwarded-Port %[dst_port]
	http-request set-header X-Forwarded-Proto https if { ssl_fc }
	http-request set-header X-Forwarded-Proto http if !{ ssl_fc }
	http-request set-header X-Forwarded-Proto-Version h2 if { ssl_fc_alpn -i h2 }
	http-request set-header X-Forwarded-Proto-Version h3 if { ssl_fc_alpn -i h3 }

	#*-------------------------*#
	#      request: denies      #
	#*-------------------------*#
	# deny if host header is restricted by ip
	http-request deny deny_status 403 if { req.hdr(host),lower,map_dom(/usr/local/etc/haproxy/maps/domain_admin_access.map) -m found } !{ src,map_ip(/usr/local/etc/haproxy/maps/admin_ip_list.map) -m found } !{ src,map_ip(/usr/local/etc/haproxy/maps/admin_ip_list_unrestricted.map) -m found }
	# deny if host header is not found in domain to backend or in redirects
	http-request deny deny_status 404 if !{ req.hdr(host),lower,map_dom(/usr/local/etc/haproxy/maps/domain_to_backend.map) -m found } !{ req.hdr(host),lower,map_dom(/usr/local/etc/haproxy/maps/redirects.map) -m found }

	#*-------------------------*#
	#      request: auth        #
	#*-------------------------*#
	http-request auth realm AdminAccess if auth_required !auth_required_unrestricted !auth_ok

	#*-------------------------*#
	#    request: redirects     #
	#*-------------------------*#
	# redirect if request path is found in redirects.map file
	http-request redirect location %[req.hdr(host),lower,map_str(/usr/local/etc/haproxy/maps/redirects.map)] code 302 if { req.hdr(host),lower,map_str(/usr/local/etc/haproxy/maps/redirects.map) -m found }

	# redirect if not https and not match acls
	redirect scheme https if !{ ssl_fc } !unifi_no_ssl_redirect

	#*-------------------------*#
	#   response: set-header    #
	#*-------------------------*#
	# add the HSTS header if not exists
	http-response set-header Strict-Transport-Security "max-age=31536000; preload;" if { ssl_fc }
	# add the response version header
	http-response set-header Via "%[res.ver] haproxy"
	# add a xss-protection header to protect against xss exploits
	http-response set-header X-XSS-Protection "1; mode=block"
	# add a x-content-type-options to protect against that it isn't possible to change the content type
	http-response set-header X-Content-Type-Options "nosniff"
	# add a referrer-policy to protect against
	acl security_referrer_policy res.hdr(Referrer-Policy) -m found
	http-response set-header Referrer-Policy "same-origin" if !security_referrer_policy
	# add a permissions-policy

	# add a content-security-policy to protect against iframes and allow own domains
	http-response set-header Content-Security-Policy "frame-ancestors 'self' https://*.hidden.se;"

	# set cache control for short lived static files
	#http-response set-header Cache-Control no-cache,max-age=600 if cache_short_static_file
	# # set cache control for long lived static files
	#http-response set-header Cache-Control public,max-age=3600 if cache_long_static_file

	http-response set-header X-Cache-Status HIT if !{ srv_id -m found }
	http-response set-header X-Cache-Status MISS if { srv_id -m found }

	# 'Alt-Svc' header invites client to switch to the QUIC protocol
  	# Max age (ma) is set to 15 minutes (900 seconds), but
  	# can be increased once verified working as expected
  	http-response set-header alt-svc "h3=\":443\";ma=900;" if { ssl_fc }


	#*-------------------------*#
	#   response: del-header    #
	#*-------------------------*#
	# remove headers we dont want to show to client
	http-response del-header Server
	http-response del-header X-Powered-By
	http-response del-header X-AspNet-Version

	#*-------------------------------*#
	# after-response: replace-header  #
	#*-------------------------------*#

	#*-------------------------*#
	#      backend: lookup      #
	#*-------------------------*#
	# # use backend letsencrypt if request is acme challenge
	# use_backend backend_letsencrypt if letsencrypt_acl

	# use backend haproxy_status if request match status page
	#use_backend backend_haproxy_status if status_page { src,map_ip(/usr/local/etc/haproxy/maps/admin_ip_list.map) -m found }

	# use correct backend servers from domain_to_backend.map file, default to no-match
	use_backend %[req.hdr(host),lower,map_dom(/usr/local/etc/haproxy/maps/domain_to_backend.map,no-match)]

#-----------------------------------#
#            backends               #
#-----------------------------------#
backend <removed>



# return 404 page not found
backend no-match
	http-request deny deny_status 404
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.8.0-fdd8154 2023/05/31 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2028.
Known bugs: http://www.haproxy.org/bugs/bugs-2.8.0.html
Running on: Linux 5.15.0-73-generic #80-Ubuntu SMP Mon May 15 15:18:26 UTC 2023 x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_LINUX_TPROXY=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_TFO=1 USE_QUIC=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 -BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION +QUIC +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=8).
Built with OpenSSL version : OpenSSL 1.1.1t+quic  7 Feb 2023
Running on OpenSSL version : OpenSSL 1.1.1t+quic  7 Feb 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.1 20220924

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

**Comment by Darlelet** [comment]
Very sorry, this regression has been introduced with 730b9836a6ff95894fd49ed3ba32c8b80693281f

If the option is present in the frontend but missing from the backend it is assumed `!= if-none`, which conflicts with previous behavior.

Unfortunately the `ifnone-forwardfor.vtc` regtest did not cover this one, I will add this testcase to prevent any regression in the future.

The below patch should help:
```diff
diff --git a/src/http_ext.c b/src/http_ext.c
index d1bae9c2a..9d70086c7 100644
--- a/src/http_ext.c
+++ b/src/http_ext.c
@@ -790,8 +790,8 @@ int http_handle_xff_header(struct stream *s, struct channel *req)
 		struct http_hdr_ctx ctx = { .blk = NULL };
 		struct ist hdr = ((b_xff) ? b_xff->hdr_name : f_xff->hdr_name);
 
-		if (f_xff && f_xff->mode == HTTP_XFF_IFNONE &&
-		    b_xff && b_xff->mode == HTTP_XFF_IFNONE &&
+		if ((!f_xff || f_xff->mode == HTTP_XFF_IFNONE) &&
+		    (!b_xff || b_xff->mode == HTTP_XFF_IFNONE) &&
 		    http_find_header(htx, hdr, &ctx, 0)) {
 			/* The header is set to be added only if none is present
 			 * and we found it, so don't do anything.
```

As a workaround to wait for the patch, you will need to explicit `option forwardfor if-none` in the backend as well

**Comment by Darlelet** [comment]
The doc is quite ambiguous as well on this point:
```
In the case of the "if-none" argument, if at least one of
the frontend or the backend does not specify it, it wants the addition to be
mandatory, so it wins.
```

**Comment by robinostlund** [comment]
No worries, great that you where able to find it that quickly 😄  thanks for your help

**Comment by capflam** [comment]
It should be fixed. Thanks !

**Comment by capflam** [comment]
The fix was backported to 2.8. Thanks !
